common/task/sequentialTaskProcessor.go

```diff
@@ -0,0 +1,197 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package task
+
+import (
+	"sort"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/uber-common/bark"
+	"github.com/uber/cadence/common"
+)
+
+type (
+	sequentialTaskProcessorImpl struct {
+		status       int32
+		shutdownChan chan struct{}
+		waitGroup    sync.WaitGroup
+
+		coroutineSize       int
+		coroutineTaskQueues map[int]chan SequentialTask
+		logger              bark.Logger
+	}
+
+	// SequentialTasks slice of SequentialTask
+	SequentialTasks []SequentialTask
+)
+
+const (
+	coroutineTaskQueueSize           = 100
+	coroutineTaskProcessingBatchSize = 100
+)
+
+// NewSequentialTaskProcessor create a new sequential tasks processor
+func NewSequentialTaskProcessor(coroutineSize int, logger bark.Logger) SequentialTaskProcessor {
```

Comment:
this seems like a multi-queue version of the scanner executor in #1602.. I have a feeling you can build this on top of that.. That way we do not have two different queuing and processing implementations doing the similar thing..

Reply:
Actually, that may not be possible, but maybe worth you taking a look and see if you see similarities